crates/telio-wg/src/wg.rs

```diff
@@ -544,24 +592,76 @@ impl State {
             } else {
                 return false;
             };
-            if !try_send(PeerState::Connecting, peer.clone()).await {
+
+            if !try_send(PeerState::Connecting, None, peer.clone()).await {
                 return false;
             }
 
-            if peer.is_connected() && !try_send(PeerState::Connected, peer.clone()).await {
+            if peer.is_connected()
+                && !try_send(PeerState::Connected, Some(LinkState::Up), peer.clone()).await
+            {
                 return false;
             }
         }
 
+        let current_link_state = |peer| {
+            if matches!(self.no_link_detection, NoLinkDetection::Disabled) {
+                return None;
+            }
+
+            let new_rx = to.peers.get(peer).and_then(|p| p.rx_bytes);
+            let old_rx = self.last_rx_timestamp.get(peer).map(|ts| ts.1);
+
+            if new_rx != old_rx {
+                if new_rx.is_some() {
+                    Some(LinkState::Up)
+                } else {
+                    None
+                }
+            } else {
+                let configurable_rtt = match self.no_link_detection {
+                    NoLinkDetection::Disabled => DEFAULT_CONFIGURABLE_RTT,
+                    NoLinkDetection::Enabled { rtt } => rtt,
+                };
+                self.last_rx_timestamp.get(peer).map(|ts| {
+                    if ts.0.elapsed() > WG_KEEPALIVE + configurable_rtt {
+                        LinkState::Down
+                    } else {
+                        LinkState::Up
+                    }
+                })
+            }
+        };
+
         // Check for updates, and notify
         for key in &diff_keys.update_keys {
             if let (Some(old), Some(new)) = (from.peers.get(key), to.peers.get(key)) {
                 let old_state = old.state();
                 let new_state = new.state();
 
+                let old_link_state = self.last_link_state.get(key).copied();
+                let new_link_state = current_link_state(key);
+
+                let should_notify_link_state_change = |new_state, old, new: Option<LinkState>| {
+                    if matches!(self.no_link_detection, NoLinkDetection::Disabled) {
+                        return false;
+                    }
+
+                    new_state == PeerState::Connected && old != new && new.is_some()
```

Comment:
This is correct logic, however if everything is as we expect, `new` should always be `Some(T)` when `new_state == PeerState::Connected`. For this I suggest of removing this additional check and adding a warning if that's the case instead. This way we could catch that it's happening and fix it in the future.